Send also workflows as part of unified jobs and send all changes to jobs

Ladas opened the pull request:
##### SUMMARY
Workflows do not have a record in main_job, therefore the JOIN
was ignoring those. We need to do LEFT JOIN to include also
workflows.

It can take several hours for a job to go from pending to successful/failed state and we need to also send the job with a changed state, otherwise the analytics will be incorrect.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - Analytics

##### FIXES
https://github.com/ansible/awx/issues/6200

Partially fixing: https://github.com/ansible/awx/issues/6201 (sending workflows as part of unified jobs)

##### Works with current analytics processor?

Currently the processor will break, since it requires Org name and ID as NOT NULL. I will limit this constraints for only template jobs. (will insert PR here)


##### TODO
- [x] figure out if we want to add organization relation from workflow_template (for some reason workflow_job misses it)
- [x] adjust analytics processor to deal with missing org for workflows https://github.com/RedHatInsights/tower-analytics-backend/pull/262/commits/793721f624e5fcd07c2089a158f76935e4c20ad7

Ladas commented:
cc @rooftopcellist @wenottingham @benthomasson

Ladas committed:
Send also workflows as part of unified jobs

Workflows do not have a record in main_job, therefore the JOIN
was ignoring those. We need to do LEFT JOIN to include also
workflows.

It also seems like we are not able to get a link to organizations
from workflows? When looking at:
<tower_url>#/organizations?organization_search=page_size:20;order_by:name

We don't seem to list a relation to workflows. Is it possible to get it from
somewhere?

Ladas committed:
Use modified to check if job should be sent to analytics

It can take several hours for a job to go from pending to
successful/failed state and we need to also send the job with
a changed state, otherwise the analytics will be incorrect.

ryanpetrello commented:
We may want to hold off on this until @AlanCoding's change here lands: https://github.com/ansible/awx/pull/5726/

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -275,11 +275,11 @@ def _copy_table(table, query, path):
-                                 JOIN main_organization ON main_organization.id = main_project.organization_id
-                                 WHERE main_unifiedjob.created > {} 
+                                 LEFT JOIN main_project ON main_project.unifiedjobtemplate_ptr_id = main_job.project_id
+                                 LEFT JOIN main_organization ON main_organization.id = main_project.organization_id
```
This join is going to change after @AlanCoding's PR merges.

Ladas replied:
cool, I am fine fixing this another way. Just keep in mind this was basically filtering out everything except jobs. But we want other stuff like `workflowapproval`, `worflowjob`, `projectupdate` etc. , so we'll need the left join, since not everything has org associated

Also @AlanCoding is your PR going to be backported? We need this backported to all versions sending data to analytics, so we get the workflow and other data.

AlanCoding replied:
no, we do not plan on backporting. that would be very complicated

Ladas replied:
Hm, ok, so can we merge this first? Or it doesn't matter for easier backport?

ryanpetrello replied:
We should wait for @AlanCoding's organization schema to land in `devel` before changing this in `devel`.

If you want to open a PR _like this_ in a backport branch, we'll have to carry a different implementation for that database schema anyways.

Ladas replied:
sounds good, I'll backport and wait for @AlanCoding here

AlanCoding commented:
> It also seems like we are not able to get a link to organizations
from workflows? When looking at:
<tower_url>#/organizations?organization_search=page_size:20;order_by:name

I don't follow that URL. If you were searching from `/api/v2/unified_jobs/` or `/api/v2/workflow_jobs/`, or the corresponding template endpoints, and using `organization`, then yes, I see why that wouldn't work, and that will change with the other PR I have open because it moves the field from the workflow tables to the unified tables

Ladas commented:
@AlanCoding the tower url was just supposed to show that on the organization page, we don't show any links to workflows. And as I was scanning the db, we only store organization relation in workflow template, not workflow_job.

ryanpetrello commented:
@Ladas we've merged @AlanCoding's work - this is probably ready for a rebase and another look now.

AlanCoding commented:
why was the created filter changed to `modified`? I think those were both indexed the same, but this is confusing. What do you think that `modified` represents? And how does that differ from `finished`? I'm sorry, I'm still groking what data it is that you are trying to obtain.

Ladas commented:
@AlanCoding the use of modified is described in the issue (I suppose finished can be used also). If we use created, then finished/modified can be hours later and if we'd already sent the payload, we wouldn't send the job again. We need to send the job again if it's finished (or change occurred) so the analytics contains the terminal state.

AlanCoding commented:
I may not have access to the issue. But `finished` is indexed

https://github.com/ansible/awx/blob/4a5e596653cf5ac144e92a8481f10f2d26b90e96/awx/main/models/unified_jobs.py#L631-L637

can't say the same for modified. So it seems like that'll work as a drop-in replacement here.

And the organization field should be on the unified jobs table now, but I think there are more problems you're trying to solve here which I'm not fully aware of.

Ladas commented:
@AlanCoding I see. Although finished doesn't have `NOT NULL` constraint. And from what I see, it is being set only when really finished. So it will not catch jobs stuck in pending etc. (so things we are trying to detect and alert upon)

So it looks like we have 2 paths:

1) using `modified`

GOOD: we'll catch all changes
BAD: we'll always have to do a sequential scan of unified jobs

2) using both `created` and `finished`

GOOD: we'll use indexes
BAD: while we should catch all jobs, the state it is stuck on doesn't have to be correct

------

Looks like 2) is good enough, we just need take in account that when `finished` is NULL, we might not have the right status. (and e.g. change the alert for `too many pending jobs` to `too many unfinished jobs`)


----

For the orgs, in analytics I expect them to be there for `job` type. So it's ok that the rest of unified jobs don't have it. But we should send all unified jobs for analytics except what is explicitly filtered out in where condition.

Ladas committed:
Use indexed timestamps

Use created and finished, which are indexed, to try to fetch all
states of jobs. If job is not finished, we might not get the
right terminal status, but that should be ok for now.

Ladas commented:
@AlanCoding ok I changed the condition to `(main_unifiedjob.created > {0} OR main_unifiedjob.finished > {0})` and I've added the left join, so the unified jobs without org are not filtered out.

Can you review also the backports please? https://github.com/ansible/tower/pull/4180 and https://github.com/ansible/tower/pull/4179


Btw. we should probably add index to modified column, so we can get the 100% correct data to analytics. (even if that can't be backported)

wenottingham commented:
What still needs done here? Just a review?

cc @AlanCoding @rooftopcellist

ryanpetrello commented:
@wenottingham likely that and verification from somebody.

chrismeyersfsu committed:
test analytics table output

* unified_jobs output should include derived jobs i.e. project update,
inventory update, job
* This PR adds a test to ensure that.

chrismeyersfsu commented:
@Ladas I submitted a PR against your PR. https://github.com/Ladas/awx/pull/1/files incorporate that test and I'm good with merging this PR  :+1:

Ladas commented:
@chrismeyersfsu thank you

btw. this PR's condition works for backports nicely. But as extension for the next major AWX release, the best would be if we could add index on `modified` column and use that instead of `created+finished`. (then we could track exact state of a job, in a case it gets stuck in a non terminal state)

ryanpetrello commented:
@Ladas the CI failures here suggest to me that this branch needs to be rebased against `devel`.

Ladas commented:
@ryanpetrello thanks for the hint, rebased

wenottingham commented:
@ryanpetrello ok for mergeit with @chrismeyersfsu's approval?

kdelee commented:
linking https://github.com/ansible/awx/issues/5931
